tornettools/generate.py

```diff
@@ -220,26 +233,66 @@ def __server(args, network, server):
     # tgen starts at the end of shadow's "bootstrap" phase
     process["start_time"] = BOOTSTRAP_LENGTH_SECONDS
 
+    if server['is_hs_server']:
+        # this is an onion service, so tgen should only listen on localhost
+        process["environment"] = "TGENIP=127.0.0.1"
+
     host["processes"].append(process)
 
+    if server['is_hs_server']:
+        # prepare the hostname and hs_ed25519_secret_key files for the onion service
+        hosts_prefix = "{}/{}/{}".format(args.prefix, SHADOW_TEMPLATE_PATH, SHADOW_HOSTS_PATH)
+        server_prefix = "{}/{}".format(hosts_prefix, server['name'])
+        hs_prefix = "{}/{}".format(server_prefix, TOR_ONIONSERVICE_DIR)
+
+        if not os.path.exists(hs_prefix):
+            os.makedirs(hs_prefix, 0o700)
+
+        with open("{}/{}".format(hs_prefix, 'hostname'), 'w') as outf:
+            outf.write(server['hs_hostname']+'\n')
+        with open("{}/{}".format(hs_prefix, 'hs_ed25519_secret_key'), 'wb') as outf:
+            outf.write(b"== ed25519v1-secret: type0 ==\x00\x00\x00" + base64.b64decode(server['hs_ed25519_secret_key']))
+
+        # tor process for the onion service
+        process = {}
+        process["path"] = "{}/bin/tor".format(SHADOW_INSTALL_PREFIX)
+        process["args"] = __format_tor_args(None)
+        process["start_time"] = BOOTSTRAP_LENGTH_SECONDS-60 # start before boostrapping ends
```

Comment:
Careful here: if `BOOTSTRAP_LENGTH_SECONDS` ever gets set to a value less than or equal to 60 seconds, we'll have errors in shadow (trying to start at a non-positive time).

Update: I see this happens in other parts of the code too. We could also just add a comment to the definition of `BOOTSTRAP_LENGTH_SECONDS` saying that we don't want it to be less than 61.

Reply:
How about changing them all to a `max(0, BOOTSTRAP_LENGTH_SECONDS-60)`?

Reply:
Don't we need it to be at least 1? Or can we start at time t=0?

Reply:
As far as I know it's fine to start at t=0 (it works for me), but most of our tests start at t=1, so maybe there's some issue (or was an issue at some point) I don't know about if processes start at t=0? But I'll make this `max(1, BOOTSTRAP_LENGTH_SECONDS-60)` just in case.